### DISCUSSION TITLE
Lab3 bram address problem

### DISCUSSION THREAD

**Discussion by v00000v** [bol-edu/HLS-SOC-Discussions#87] [Q&A]
有兩個問題想請教各位:
1. 請問在bram中，當對RAM進行讀寫時，address需右移兩個bit的原因是什麼呢?

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/123727462/cc4cafb6-eb76-4180-b0e6-bf3a55c8d677)

2. 當testbench在傳讀寫是否需要每次+4，才不會重複傳到同一個地址呢?

**Comment by edwardchang9** [comment]
同學好：
個人淺見供參，歡迎討論
1. 因為bram 資料大小是8bit，然後bram.v裡面儲存資料的方式是以word為單位的矩陣，因此addtess要右移兩個bit 才會對應到bram 裡面要存哪個word的矩陣index
2. 是的

**Reply by TINGxTING** [reply]
同學你好:
可以請問為何bram資料大小要設計成8bit嗎?
是因為byte address的關係，所以大小是8bit嗎?
謝謝你!

**Reply by edwardchang9** [reply]
沒錯!

**Reply by kevin861222** [reply]
" >>2 "可以想成是 offset 的概念，因為在這個bram的設定中，一次讀取32bits = 8*2^(offset) ， offset = 2

**Comment by DanielDu9453** [comment]
1. 我認為這邊的address應該是在模擬記憶體中byte address的格式，每一個address都指向1byte的資料。因為記憶體每次讀寫都是4個bytes，所以才要將address位移，以address/4的位置為起點開始寫入4個bytes，這樣的address叫做word address。詳細的資料可以查詢memory alignment及word address

2. 是的，每筆32bits的資料都要寫在address能夠被4整除的位置

如果以上回答有誤歡迎糾正

**Comment by michael61112** [comment]
想補充請問一下各位，我看範例bram11.v裡面，RAM宣告每一個address內的大小是32bit，並不是1btye，
如此我搞不懂為何還需要一次shift四個address? 
=> 還是這就是要留給寫作業的同學改回8bit的宣告? 
=> reg [7:0] RAM[0:10];

![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/9332860/b4b010c9-b7f7-42f1-b6df-2d4967209558)

**Reply by edwardchang9** [reply]
首先，這個bram是11個word大小，一個word是4byte，所以絕對不可能改成reg [7:0] RAM[0:10];，如果改成這樣大小只剩11bytes
關於你的問題，送進bram的address是以byte為單位的address，但是對bram來說，他的address是以word為單位的
因此在address的轉換上才需要shift 2 bit

如果以上回答有誤歡迎糾正
